chore(next.clayui.com): add more docs on using Modal (#2591)
liferay/clay: next.clayui.com/content/docs/components/modal.mdx

<Modal />

### useModal hook

The `useModal` [hook](https://reactjs.org/docs/hooks-intro.html) is required to create a Modal component in Clay, it is responsible for communicating with ClayModal and controlling states to make animations, this is necessary because Modal needs animation when opening and closing so we need to delay calls of the `onClose` so you can safely control Modal's visibility or do other things.

`useModal` returns a signature with `observer` and `onClose`:

> - **observer** - Observer is the communication channel that connects ClayModal with `useModal`, you can call it a tunnel.

Reviewer: Not sure the "you can call it a tunnel." part is helpful, it might add to the confusion.

Author: Hmm yeah, I will remove it.

- **onClose** - Callback is called after Modal's close animation.

```jsx
const { observer, onClose } = useModal({
    onClose: () => setVisible(false)
});
```

<div class="clay-site-alert alert alert-warning">

>     To avoid future problems, do not use the information or change the values ​​of `observer`, it is considered a bad practice.

Reviewer: > "do not use the information"

Is this accurate? Makes sense to discourage the changing of the values, but reading the values doesn't seem like a bad pattern.

Author: I didn't want to encourage reading these values, because we can change the information contained in `observer`, so I've been thinking of providing an API to` useModal` to address the need to know when Modal is really visible after the animation is done.

Reviewer: yeah thats a good idea to add an API, I don't think we should dissuade people from reading the values unless we have another means of reading that information

Author: Yeah, we can add this and wait for the release of Clay 3.0 officially, I'll send a PR later adding this API. But still I would like to keep this information.

Reviewer: sounds good. feel free to merge

</div>

> To prevent Modal from closing before animation never set its state directly in button events to close the modal instead call the `onClose` method returned by `useModal`.

Reviewer: ```suggestion
To prevent Modal from closing before animation always use the `onClose` method returned by `useModal`.
```

I think this should be clear and info info to convey that we want them to use the hook.

```diff
<ClayModal.Footer
-   last={<ClayButton onClick={() => setVisible(false)}>{"Primary"}</ClayButton>}
+   last={<ClayButton onClick={onClick}>{"Primary"}</ClayButton>}
/>
```

#### Usage

> `UseModal` is a hook so it implies that you cannot use ClayModal in a class component, to use you can adopt some strategies:

Reviewer: ```suggestion
`UseModal` is a hook so it implies that you cannot use ClayModal directly in a class component. If you'd like to use it in a class component, try the recommended strategy below.
```

> - Wrap ClayModal and `useModal` into one component.

Reviewer: ```suggestion
- Wrap ClayModal and `useModal` in a function component.`
```

> - Create your modal expert on a functional component.

Reviewer: Not sure this makes sense, we can probably remove it and just include the one strategy with the code example.

Author: I will remove this and just leave the example below if we see more cases we can add them.

    const { observer, onClose: close } = useModal({
        onClose: onVisible,
    });

    if (!visible) {
        return;
    }

    return (
        <ClayModal
            observer={observer}
            spritemap={spritemap}
        >
            <ClayModal.Header>{title}</ClayModal.Header>
            <ClayModal.Body>
                {body}
            </ClayModal.Body>
            <ClayModal.Footer
                last={<ClayButton onClick={close}>{"Close"}</ClayButton>}
            />
        </ClayModal>
    );
};

// This is pseudocode
class Page extends Component {
    render() {
        return (
            <>
                <ModalWithDialog visible={this.state.visible} onClose={() => this.setVisible(false)} title={...} body={...} />
                <ClayButton displayType="primary" onClick={() => this.setVisible(true)}>
                    {"Open modal"}
                </ClayButton>
            <>
        );
    }
}
```

`useModal` may be lost if you raise the visible condition above the hook, if Modal happens to be unmounted at the same time as the component. Ideally the hook is not unmounted every time Modal is opened and closed.

### Provider

> When your application uses a lot of modals, it may be bad to be creating multiple modes for your application, you can use `<ClayModalProvider />` to have only one open Modal on the screen and the component centered on your application, you can invoke the modal from the context through the API.

Reviewer: ```suggestion
When your application uses a lot of modals, you can use `<ClayModalProvider />` to have only one open Modal on the screen and the component centered on your application, you can invoke the modal from the context through the API. It is not recommended to have multiple modals open at the same time.
```
